s3/src/main/java/io/aiven/kafka/tiered/storage/s3/S3ClientWrapper.java

```diff
@@ -0,0 +1,251 @@
+    private final S3EncryptionKeyProvider s3EncryptionKeyProvider;
+
+    public S3ClientWrapper(final S3RemoteStorageManagerConfig config,
+                           final AmazonS3 s3Client,
+                           final S3EncryptionKeyProvider s3EncryptionKeyProvider) {
+        this.config = config;
+        this.s3Client = s3Client;
+        this.s3EncryptionKeyProvider = s3EncryptionKeyProvider;
+    }
+
+    public InputStream fetchLogFile(final RemoteLogSegmentMetadata remoteLogSegmentMetadata,
+                                    final int startPosition,
+                                    final int endPosition) throws RemoteStorageException {
+        final String logFileKey = getFileKey(remoteLogSegmentMetadata, LOG_FILE_SUFFIX);
+
+        final int firstPart = startPosition / config.s3StorageUploadPartSize();
+        final int lastPart;
+        if (remoteLogSegmentMetadata.segmentSizeInBytes() % config.s3StorageUploadPartSize() == 0) {
+            lastPart = endPosition / config.s3StorageUploadPartSize();
+        } else {
+            lastPart = endPosition / config.s3StorageUploadPartSize() + 1;
+        }
+        final String metadataFileKey = getFileKey(remoteLogSegmentMetadata, METADATA_FILE_SUFFIX);
+        final SecretKey encryptionKey = s3EncryptionKeyProvider.createOrRestoreEncryptionKey(metadataFileKey);
+        final CryptoIOProvider cryptoIOProvider = new CryptoIOProvider(encryptionKey, config.ioBufferSize());
+        final ArrayList<InputStream> inputStreams = new ArrayList<>();
+        for (int i = firstPart; i < lastPart - 1; i++) {
+            final String key = logFileKey + "_" + i;
+            try {
+                final GetObjectRequest getObjectRequest = new GetObjectRequest(config.s3BucketName(), key);
+                final S3Object s3Object = s3Client.getObject(getObjectRequest);
+                inputStreams.add(cryptoIOProvider.decryptAndDecompress(s3Object.getObjectContent()));
+            } catch (final Exception e) {
+                throw new RemoteStorageException(String.format("Error fetching log segment data from %s", key), e);
+            }
+        }
+        final int end = endPosition % config.s3StorageUploadPartSize();
+        final String key = logFileKey + "_" + (lastPart - 1);
+        try {
+            final GetObjectRequest getObjectRequest = new GetObjectRequest(config.s3BucketName(), key);
+            final S3Object s3Object = s3Client.getObject(getObjectRequest);
+            final BoundedInputStream boundedInputStream =
+                    new BoundedInputStream(cryptoIOProvider.decryptAndDecompress(s3Object.getObjectContent()), end);
+            inputStreams.add(boundedInputStream);
+        } catch (final IOException e) {
+            throw new RemoteStorageException(String.format("Error fetching log segment data from %s", key), e);
+        }
+        return new SequenceInputStream(Collections.enumeration(inputStreams));
```

Comment:
What I do not like about this solution is that you store all streams/bytes in the memory. Is it possible to use it differently e.g. just create an iterator that will build `InputStream` on demand?
And yet another question `BoundedInpytStream` was introduced for `ServletInputStreams` since they could be blocked due to `Content-Length` why do we use it? What is the diff compare to `ByteArrayInputStream`?

Reply:
As far as I understand the data will be fetched from S3 only during the reading of the result stream since in the javadoc of `S3Object#getObjectContent` specified that stream will be created on demand(AFAIU):

> Note: The method is a simple getter and does not actually create a stream. If you retrieve an S3Object, you should close this input stream as soon as possible, because the object contents aren't buffered in memory and stream directly from Amazon S3. 

As for `BoundedInputStream` it seems to be suitable because we can limit the content up to specified length. That's what we actually need - read up to end position. And not eagerly like it would be with `ByteArrayInputStream`because it's constructor accepts only byte arrays but on demand later. But maybe I'm missing something.

Reply:
Yes, you are right to use `ByteArrayInputStream` we need to read everything into memory. I still dout that we need `BoundedInputStream`. Lets see how it will behave